src/host/multiple_ledger.h

```diff
@@ -0,0 +1,815 @@
+    {
+      if (idx == 0)
+      {
+        return nullptr;
+      }
+
+      // First, check if the file is in the list of files open for writing
+      auto f = std::upper_bound(
+        files.rbegin(),
+        files.rend(),
+        idx,
+        [](size_t idx, const std::shared_ptr<LedgerFile>& f) {
+          return idx >= f->get_start_idx();
+        });
+
+      if (f != files.rend())
+      {
+        return *f;
+      }
+
+      // Otherwise, return file from read cache
+      return get_file_from_cache(idx);
+    }
+
+    std::shared_ptr<LedgerFile> get_latest_file() const
+    {
+      if (files.empty())
+      {
+        return nullptr;
+      }
+      return *(files.rbegin()++);
+    }
+
+  public:
+    MultipleLedger(
+      const std::string& ledger_dir,
+      ringbuffer::AbstractWriterFactory& writer_factory,
+      size_t chunk_threshold,
+      size_t max_read_cache_size = max_read_cache_size_default) :
+      ledger_dir(ledger_dir),
+      chunk_threshold(chunk_threshold),
+      to_enclave(writer_factory.create_writer_to_inside()),
+      max_read_cache_size(max_read_cache_size)
+    {
+      if (chunk_threshold == 0 || chunk_threshold > max_chunk_threshold_size)
+      {
+        throw std::logic_error(fmt::format(
+          "Error: Ledger chunk threshold should be between 1-{}",
+          max_chunk_threshold_size));
+      }
+
+      if (fs::is_directory(ledger_dir))
+      {
+        // If the ledger directory exists, recover ledger files from it
+        for (auto const& f : fs::directory_iterator(ledger_dir))
+        {
+          files.push_back(
+            std::make_shared<LedgerFile>(ledger_dir, f.path().filename()));
+        }
+
+        files.sort([](
+                     const std::shared_ptr<LedgerFile>& a,
+                     const std::shared_ptr<LedgerFile>& b) {
+          return a->get_last_idx() < b->get_last_idx();
+        });
+
+        last_idx = get_latest_file()->get_last_idx();
+
+        for (auto f = files.begin(); f != files.end();)
+        {
+          if ((*f)->is_compacted())
+          {
+            compacted_idx = (*f)->get_last_idx();
+            auto f_ = f;
+            f++;
+            files.erase(f_);
+          }
+          else
+          {
+            f++;
+          }
+        }
+
+        // Continue writing at the end of last file only if that file is not
+        // complete
+        if (files.size() > 0 && !files.back()->is_complete())
+        {
+          require_new_file = false;
+        }
+        else
+        {
+          require_new_file = true;
+        }
+      }
+      else
+      {
+        if (!fs::create_directory(ledger_dir))
+        {
+          throw std::logic_error(fmt::format(
+            "Error: Could not create ledger directory: {}", ledger_dir));
+        }
+        require_new_file = true;
+      }
+    }
+
+    MultipleLedger(const MultipleLedger& that) = delete;
+
+    const std::vector<uint8_t> read_entry(size_t idx)
```

Comment:
Nitpick: I'd prefer `std::optional<std:vector<uint8_t>>` here, so we can distinguish missing entries from present-but-empty entries (sure the latter is probably invalid, but that's not for us to decide). Also there's little point in returning const values.

Reply:
Agreed, although in practice it's hard to imagine a valid use case for empty entries.

Reply:
Good point. Done now.

Reply:
The case for empty entries was actually used when sending raft heartbeats. It was calling `read_framed_entries(from = N+1, to = 6)` which would return an empty vector. I've adapted the code slightly in `node_connections.h`.